tools/checkrelease.sh

```diff
@@ -19,114 +19,218 @@
-    echo "   Given release name or release full URL, downloads all files in"
-    echo "   in that directory (which should include nuttx and nuttx-apps"
-    echo "   sha512, asc, and tar.gz files), checks the release SHA512 and GPG "
-    echo "   signatures, and checks the unpacked directories for required "
-    echo "   files. Creates a temporary directory to do its work in."
+    echo "Usage: $0 [--verbose] [--url <URL-of-release-dir>] [--release <name-of-release] [--dir <path-to-directory>] [--tempdir <path-to-directory>]"
+    echo "   Given release full URL, release name, or a local directory, downloads or copies"
+    echo "   all files in that directory (which for a release should include nuttx and nuttx-apps, sha512, "
+    echo "   asc, and tar.gz files), checks the release SHA512 and GPG signatures, checks the unpacked "
+    echo "   directories for required files, and tries to build NuttX for sim:nsh."
+    echo
+    echo "   If tempdir is specified, it will be removed and recreated; if it is not specified, /tmp/nuttx-checkrelease"
+    echo "   is used."
     echo
-    echo "   nuttx and nuttx-apps tar.gz files are left in the current"
-    echo "   directory."
+    echo "   If --url or --release are given, nuttx and nuttx-apps tar.gz files are left in /tmp for the user to "
+    echo "   build for their platform."
     echo 
     echo "Examples:"
     echo
-    echo "  $0 9.1.0-RC1"
-    echo "  $0 https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC1/"
+    echo "  $0 --release 9.1.0-RC1"
+    echo "  $0 --url https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC1/"
+    echo "  $0 --dir ./some-dir-that-has-nuttx-and-apps"
     echo
 }
 
-if [ "-h" == "$1" ]; then
-    usage
-    exit 0
-fi
+POSITIONAL=()
+while [[ $# -gt 0 ]]
+do
+key="$1"
 
-if [ -z "$1" ]; then
+URL=""
+DIRECTORY=""
+
+case $key in
+    -U|--url)
+    URL="$2"
+    shift # past argument
+    shift # past value
+    ;;
+    -R|--release)
+    RELEASE="$2"
+    URL="$BASE_URL/$RELEASE/"
+    shift # past argument
+    shift # past value
+    ;;
+    -D|--dir)
+    DIRECTORY="$(readlink -f $2)"
+    shift # past argument
+    shift # past value
+    ;;
+    -T|--tempdir)
+    TEMPDIR="$(readlink -f $2)"
+    shift # past argument
+    shift # past value
+    ;;
+    -h|--help)
+    shift # past argument
     usage
     exit 0
-fi
+    ;;
+    -V|--verbose)
+    VERBOSE=1
+    shift # past argument
+    shift # past value
+    ;;
+    *)    # unknown option
+    POSITIONAL+=("$1") # save it in an array for later
+    shift # past argument
+    ;;
+esac
+done
+set -- "${POSITIONAL[@]}" # restore positional parameters
 
-ARG=$1
-if [[ "$ARG" =~ ^"http".* ]]; then
-  URL="$1/"
-else
-  URL="$BASE_URL/$1/"
+if [[ (-z "$URL") && (-z "$DIRECTORY") ]]; then
+  usage
+  exit 1
 fi
 
 download_release
-check_nuttx 
-check_nuttx_apps 
+if [[ -n "$URL" ]]; then
+  check_nuttx
+  check_nuttx_apps
+else
```

Comment:
Why not directly call check_nuttx and check_nuttx_apps?

Reply:
@xiaoxiang781216 I'm not sure what you mean here— do you mean, why am I checking for the $URL? If so, it's because there's an option to check a directory on disk, and those functions assume a URL is given... and so will try to download files...

Reply:
But check_nuttx/check_nuttx_apps should work with local directory too if my understanding is correct.

Reply:
@xiaoxiang781216 The check_nuttx/check_nuttx_apps expect tar.gz files and try to decompress them. I thought about adding arguments to suppress them, but factoring the code this way seemed cleaner.

Reply:
I guess what @btashton want is check the release is good before uploading to https://dist.apache.org/repos/dist/dev/incubator/nuttx/. So the local directory should contain the same files as the website.